### PR TITLE
[BREAKING CHANGE] Integrate credentials provider with the REST client

### LINKED ISSUE #16837: Provide OAuth support fro REST API calls in the Java client

### PR DESCRIPTION
**Pull request by npepinpe**
## Description

This PR integrates the `CredentialsProvider` into the REST client, allowing it to use the built-in `OAuthCredentialsProvider`.

To do so, the interface is changed in a breaking manner to support multiple protocols (as we will continue supporting both REST and gRPC in the future).

The previous API was tightly coupled to gRPC by using its own types directly. Instead, we introduced two abstractions:

- `CredentialsApplier`: a functional interface which lets you add headers to a request.
- `StatusCode`: an interface representing a the status of a failed request.

> [!Note]
> `CredentialsApplier` could be renamed to something more like `HeaderMutator` or something like this, as you prefer.
> Similarly, `StatusCode` could be renamed to `FailedStatus` or whatever you want.
>
> Naming is hard, and I'm bad at it :)

So `applyCredentials` now takes a `CredentialsApplier` (properly implemented for REST or gRPC) instead of the gRPC specific type `Metadata`, and `shouldRetryRequest` takes in a `StatusCode` instead of `Throwable`, as we anyway only care to know if it was unauthenticated or not. We couldn't reuse the same `Throwable` because then you'd need to know about library specific types (thus increasing our public API), and we can't just use the code, since gRPC and normal HTTP have different codes.

The bulk of the changes are tests related however, primarily because the previous tests were very tightly coupled to gRPC, so that was fun to change...

The PR could be split to be smaller by having the gRPC changes first, then REST, but considering the time crunch I hope this is OK.

More QA tests will be added with the server part.

## Related issues

closes #16837 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Other teams:
If the change impacts another team an issue has been created for this team, explaining what they need to do to support this change.
- [ ] [Tasklist](https://github.com/camunda/tasklist/issues)
- [ ] [Web Modeler](https://github.com/camunda/web-modeler/issues)
- [ ] [Desktop Modeler](https://github.com/camunda/camunda-modeler/issues)
- [ ] [Optimize](https://github.com/camunda/camunda-optimize/issues)

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

File: zeebe/clients/java/src/test/java/io/camunda/zeebe/client/OAuthCredentialsProviderTest.java
Comment (npepinpe) [inline zeebe/clients/java/src/test/java/io/camunda/zeebe/client/OAuthCredentialsProviderTest.java:1]: That class is gigantic :sweat_smile: I know there's some opportunity to slim down the tests, specifically the client tests, but honestly it was a pain to do. I think it's a good future improvement though.

Comment (koevskinikola) [reply zeebe/clients/java/src/test/java/io/camunda/zeebe/client/OAuthCredentialsProviderTest.java:1]: That's a lot of refactoring! I'll create a tech debt issue for this.

File: zeebe/qa/integration-tests/src/test/java/io/camunda/zeebe/it/multitenancy/MultiTenancyOverIdentityIT.java [hunk 1]
```diff
@@ -726,7 +726,7 @@ void shouldDenyActivateJobWhenUnauthorized() {
 
       // then
       assertThat(result)
-          .failsWithin(Duration.ofSeconds(10))
+          .failsWithin(Duration.ofHours(10))
```
Comment (koevskinikola) [inline zeebe/qa/integration-tests/src/test/java/io/camunda/zeebe/it/multitenancy/MultiTenancyOverIdentityIT.java:729]: :question: Why did you change this to 10 hours? Isn't that a bit much?

**Review by koevskinikola — APPROVED** [review]
@npepinpe the changes look good, great work! :tada: 

I just have some minor feedback and questions, but they aren't impacting the correctness of the code of this PR.

File: zeebe/qa/integration-tests/src/test/java/io/camunda/zeebe/it/multitenancy/MultiTenancyOverIdentityIT.java (hunk 1 above)
Comment (npepinpe) [reply zeebe/qa/integration-tests/src/test/java/io/camunda/zeebe/it/multitenancy/MultiTenancyOverIdentityIT.java:729]: No, good catch, that was for debugging (if you use a debugger then the fails within always ends up stopping you :smile:)

File: zeebe/clients/java/src/test/java/io/camunda/zeebe/client/OAuthCredentialsProviderTest.java
Comment (npepinpe) [reply zeebe/clients/java/src/test/java/io/camunda/zeebe/client/OAuthCredentialsProviderTest.java:1]: Yeah, I think it can be further refactored, esp. the duplication of the behavior with REST and gRPC for the client. We do have QA tests I guess, but the idea was to have only client integration tests (so testing not only the OAuthCredentialsProvider but it's usage in the client), and minimize the blast radius of potential failures by removing server side bugs. But yeah, it adds more stuff, and the mocking is quite different.

File: zeebe/qa/integration-tests/src/test/java/io/camunda/zeebe/it/multitenancy/MultiTenancyOverIdentityIT.java (hunk 1 above)
Comment (npepinpe) [inline zeebe/qa/integration-tests/src/test/java/io/camunda/zeebe/it/multitenancy/MultiTenancyOverIdentityIT.java:729]:
```suggestion
          .failsWithin(Duration.ofSeconds(10))
```
